### PR TITLE
Fix for "Invalid partition key range" error during partition splits - draining child partitions on retry

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**

### CODE REVIEW

**Review by ealsur — APPROVED** [review]
The recursion is strong in this one

**Review by ealsur — APPROVED** [review]
Nice refactor catch!

File: src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala [hunk 1]
```diff
@@ -155,6 +155,60 @@ private[spark] case class CosmosDBConnection(config: Config) extends CosmosDBLog
...
 
+  private def tryHandlePartitionSplitError
+  (
+    outerError: Throwable,
+    error: DocumentClientException,
+    originalFeedOptions: FeedOptions,
+    counterSnapshot: Int
+  ) : (Int, Array[String]) = {
+    if (counterSnapshot > 1000 ||
+      error.getStatusCode() != 410) {
```
Comment (ealsur) [inline src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala:166]: only one nit, maybe it just won't ever happen, but could the `error` in the case of the Illegal exception be `null` or empty`?

File: src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala [hunk 2]
```diff
@@ -218,50 +257,7 @@ private[spark] case class CosmosDBConnection(config: Config) extends CosmosDBLog
...
-                pkRange.getParents.contains(originalFeedOptions.getPartitionKeyRangeIdInternal)) {
-
-                newChildRanges += pkRange.getId()
-                logInfo(s"Found child partition '${pkRange.getId}' for original " +
-                  s"partition '${originalFeedOptions.getPartitionKeyRangeIdInternal}'.")
-              }
-            })
-            childPKRanges = newChildRanges.toArray
-          }
+          tryHandlePartitionSplitError(outerError, error, originalFeedOptions, counter)
```
Comment (ealsur) [inline src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala]: nit: Should we be updating the value of childPKRanges and counter with the output of the call?

File: src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala:166]: No lines 253-255 prevent that - In the new connector I would use an assert to clarify - but in the old connector we haven't wired up asserts. And this method is private - so I think it is ok

File: src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala (hunk 2 above)
Comment (FabianMeiswinkel) [reply src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala]: Yes - thanks!

Comment (FabianMeiswinkel) [reply src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala]: Fixed in next iteration

**Review by ealsur — APPROVED** [review]

**Pull request merged**